File: skills/pretext/scripts/check_layout_api_sync.py
#!/usr/bin/env python3
"""Check whether skill docs cover the exports from pretext/src/layout.ts."""

from __future__ import annotations

import re
import sys
from pathlib import Path


FUNCTION_RE = re.compile(r"^export function (\w+)\(", re.MULTILINE)
TYPE_RE = re.compile(r"^export type (\w+)\b", re.MULTILINE)
HEADING_RE = re.compile(r"^### `([A-Za-z_]\w*)(?:\(|`)", re.MULTILINE)
BULLET_CODE_RE = re.compile(r"^- `([A-Za-z_]\w*)`$", re.MULTILINE)


def extract_layout_exports(layout_ts: str) -> set[str]:
    exports = set(FUNCTION_RE.findall(layout_ts))
    exports.update(TYPE_RE.findall(layout_ts))
    return exports


def extract_documented_names(markdown: str) -> set[str]:
    names = set(HEADING_RE.findall(markdown))
    names.update(BULLET_CODE_RE.findall(markdown))
    return names


def main() -> int:
    repo_root = Path(__file__).resolve().parents[3]
    layout_path = repo_root / "pretext" / "src" / "layout.ts"
    public_api_path = repo_root / "skills" / "pretext" / "reference" / "public-api.md"
    internal_exports_path = repo_root / "skills" / "pretext" / "reference" / "internal-exports.md"

    layout_exports = extract_layout_exports(layout_path.read_text(encoding="utf-8"))
    documented_names = extract_documented_names(public_api_path.read_text(encoding="utf-8"))
    documented_names.update(extract_documented_names(internal_exports_path.read_text(encoding="utf-8")))

    missing = sorted(layout_exports - documented_names)
    extra = sorted(documented_names - layout_exports)

    if not missing and not extra:
        print("Layout export docs are in sync.")
        return 0

    if missing:
        print("Missing documented exports:")
        for name in missing:
            print(f"- {name}")
    if extra:
        print("Documented names that are not layout.ts exports:")
        for name in extra:
            print(f"- {name}")
    return 1


if __name__ == "__main__":
    raise SystemExit(main())

File: skills/pretext/scripts/check-layout-api-sync.ts
#!/usr/bin/env bun
/**
 * Check whether skill docs cover the exports from pretext/src/layout.ts.
 *
 * Usage:
 *   bun run skills/pretext/scripts/check-layout-api-sync.ts
 *
 * Exit code 0 if in sync, 1 if mismatches found.
 */

import { join, resolve } from "node:path";

// ── Regex patterns (identical to the Python version) ────────────────────────

const FUNCTION_RE = /^export function (\w+)\(/gm;
const TYPE_RE = /^export type (\w+)\b/gm;
const HEADING_RE = /^### `([A-Za-z_]\w*)(?:\(|`)/gm;
const BULLET_CODE_RE = /^- `([A-Za-z_]\w*)`$/gm;

// ── Helpers ─────────────────────────────────────────────────────────────────

/** Collect all capture-group-1 matches for a global regex. */
function allMatches(re: RegExp, text: string): string[] {
  const results: string[] = [];
  // Reset lastIndex in case the regex was used before
  re.lastIndex = 0;
  let m: RegExpExecArray | null;
  while ((m = re.exec(text)) !== null) {
    results.push(m[1]);
  }
  return results;
}

function extractLayoutExports(layoutTs: string): Set<string> {
  const exports = new Set<string>(allMatches(FUNCTION_RE, layoutTs));
  for (const name of allMatches(TYPE_RE, layoutTs)) {
    exports.add(name);
  }
  return exports;
}

function extractDocumentedNames(markdown: string): Set<string> {
  const names = new Set<string>(allMatches(HEADING_RE, markdown));
  for (const name of allMatches(BULLET_CODE_RE, markdown)) {
    names.add(name);
  }
  return names;
}

// ── Main ────────────────────────────────────────────────────────────────────

async function main(): Promise<number> {
  // import.meta.dir → …/skills/pretext/scripts  →  3 parents up = repo root
  const repoRoot = resolve(import.meta.dir, "..", "..", "..");

  const layoutPath = join(repoRoot, "pretext", "src", "layout.ts");
  const publicApiPath = join(
    repoRoot,
    "skills",
    "pretext",
    "reference",
    "public-api.md",
  );
  const internalExportsPath = join(
    repoRoot,
    "skills",
    "pretext",
    "reference",
    "internal-exports.md",
  );

  // Read all files concurrently using Bun-native file reading
  const [layoutText, publicApiText, internalExportsText] = await Promise.all([
    Bun.file(layoutPath).text(),
    Bun.file(publicApiPath).text(),
    Bun.file(internalExportsPath).text(),
  ]);

  const layoutExports = extractLayoutExports(layoutText);

  const documentedNames = extractDocumentedNames(publicApiText);
  for (const name of extractDocumentedNames(internalExportsText)) {
    documentedNames.add(name);
  }

  // Set difference: missing = layoutExports − documentedNames
  const missing = [...layoutExports]
    .filter((n) => !documentedNames.has(n))
    .sort();
  // Set difference: extra = documentedNames − layoutExports
  const extra = [...documentedNames]
    .filter((n) => !layoutExports.has(n))
    .sort();

  if (missing.length === 0 && extra.length === 0) {
    console.log("Layout export docs are in sync.");
    return 0;
  }

  if (missing.length > 0) {
    console.log("Missing documented exports:");
    for (const name of missing) {
      console.log(`- ${name}`);
    }
  }

  if (extra.length > 0) {
    console.log("Documented names that are not layout.ts exports:");
    for (const name of extra) {
      console.log(`- ${name}`);
    }
  }

  return 1;
}

process.exit(await main());
